dali/python/backend_impl.cc

```diff
@@ -1064,11 +1073,26 @@ PYBIND11_MODULE(backend_impl, m) {
     .def("num_threads", &Pipeline::num_threads)
     .def("device_id", &Pipeline::device_id)
     .def("SetExternalTLInput",
-        [](Pipeline *p, const string &name, const TensorList<CPUBackend> &tl) {
-          p->SetExternalInput(name, tl);
-        })
+        [](Pipeline *p, const string &name, const TensorList<CPUBackend> &tl,
+           py::object /*cuda_stream*/) {
+          p->SetExternalInput(name, tl, 0);
+        },
+        "name"_a,
+        "list"_a,
+        "cuda_stream"_a = py::none())
+    .def("SetExternalTLInput",
+        [](Pipeline *p, const string &name, const TensorList<GPUBackend> &tl,
+           py::object cuda_stream) {
+           cudaStream_t stream = cuda_stream.is_none()
+                                 ? UserStream::Get()->GetStream(tl)
+                                 : static_cast<cudaStream_t>(ctypes_void_ptr(cuda_stream));
+          p->SetExternalInput(name, tl, stream);
+        },
+        "name"_a,
+        "list"_a,
+        "cuda_stream"_a = py::none())
     .def("SetExternalTensorInput",
-        [](Pipeline *p, const string &name, py::list list) {
+        [](Pipeline *p, const string &name, py::list list, py::object cuda_stream) {
```

Comment:
I think this could be handled by two overloads accepting std::vector<Tensor<CPUBackend>> and std::vector<Tensor<GPUBackend>> instead of `py::list list`. Than you probably don't need to resort to the magic pybind caster (at least not explicitly).

Reply:
See the comment:
```
// Note: This is a hack to get around weird casting
          // issues w/ pybind and a non-copyable type (dali::Tensor).
          // We cannot use pybind::cast<Tensor<CPUBackend>>
          // because somewhere through the chain of templates
          // pybind returns the calling template type, which
          // tries to call the deleted copy constructor for Tensor.
          // instead, we cast to a reference type and manually
          // move into the vector.
```

Reply:
I thought it's the magic things that are happening below, but if it's the reason why we accept py::list here, than fine.